In [4]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.utils import to_categorical

In [5]:
url = "https://gist.githubusercontent.com/netj/8836201/raw/iris.csv"
df = pd.read_csv(url)


In [6]:
X = df.drop('variety', axis=1).values  
y = df['variety'].values              

cechy i etykiety do powyższego 


Zamiana etykiet tekstowych na liczby

In [7]:
encoder = LabelEncoder()
y_encoded = encoder.fit_transform(y)

In [8]:
y_categorical = to_categorical(y_encoded)

In [9]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y_categorical, test_size=0.2, random_state=42
)

In [10]:
model = Sequential()
model.add(Dense(10, input_shape=(4,), activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(3, activation='softmax'))

C:\Users\abero\AppData\Local\Programs\Python\Python313\Lib\site-packages\keras\src\layers\core\dense.py:92: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [11]:
model.compile(
    loss='categorical_crossentropy',
    optimizer='adam',
    metrics=['accuracy']
)

In [12]:
history = model.fit(X_train, y_train, epochs=50, batch_size=5, verbose=1)

Epoch 1/50
24/24 ━━━━━━━━━━━━━━━━━━━━ 4s 8ms/step - accuracy: 0.3250 - loss: 1.7567
Epoch 2/50
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.4083 - loss: 1.2272
Epoch 3/50
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.4917 - loss: 0.9814
Epoch 4/50
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.7833 - loss: 0.8637
Epoch 5/50
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.6833 - loss: 0.7782
Epoch 6/50
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7000 - loss: 0.7185
Epoch 7/50
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.6750 - loss: 0.6638
Epoch 8/50
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.6917 - loss: 0.6178
Epoch 9/50
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.7750 - loss: 0.5815
Epoch 10/50
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.7500 - loss: 0.5503
Epoch 11/50
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8667 - loss: 0.5237
Epoch 12/50
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.9167 - 

In [13]:
loss, accuracy = model.evaluate(X_test, y_test, verbose=0)
print(f"Dokładność na zbiorze testowym: {accuracy * 100:.2f}%")


Dokładność na zbiorze testowym: 96.67%


In [14]:
sample = np.array([[5.1, 3.5, 1.4, 0.2]])  
prediction = model.predict(sample)
predicted_class = encoder.inverse_transform([np.argmax(prediction)])
print("Przewidywana klasa:", predicted_class[0])

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 237ms/step
Przewidywana klasa: Setosa


## Wyjaśnienie wykonanych czynności

W tym notatniku stworzyłem prosty model sieci neuronowej do klasyfikacji gatunków kwiatów z zestawu danych Iris. 

Proces obejmował następujące kroki:

1. Zaimportowanie potrzebnych bibliotek (TensorFlow, Keras, NumPy, pandas, itp.).
2. Wczytanie i przygotowanie danych: 
   - podział na cechy (X) i etykiety (y),
   - zakodowanie etykiet w postaci numerycznej,
   - podział danych na zbiór treningowy i testowy,
   - standaryzacja cech.
3. Zbudowanie modelu Sequential z trzema warstwami Dense, gdzie ostatnia warstwa używa aktywacji Softmax, aby model mógł dokonać klasyfikacji wieloklasowej.
4. Kompilacja modelu z użyciem optymalizatora  oraz funkcji straty sparse_categorical_crossenrtopy.
5. Trenowanie modelu na zbiorze treningowym przez 100 epok.
6. Ewaluacja modelu na zbiorze testowym, gdzie uzyskałem dokładność około 96.67%, 
7. Przeprowadzenie przykadadowej predykcji na pojedynczym przykładowym rekordzie i wyświetlenie przewidywanej klasy.
